components/raftstore/src/store/fsm/peer.rs

```diff
@@ -879,53 +869,12 @@ where
         self.register_raft_base_tick();
     }
 
-    fn finish_unsafe_recovery_wait_apply(&mut self) {
-        if self
-            .fsm
-            .unsafe_recovery_wait_apply_counter
-            .as_ref()
-            .unwrap()
-            .fetch_sub(1, Ordering::Relaxed)
-            == 1
-        {
-            let mut stats = StoreStats::default();
-            stats.set_store_id(self.store_id());
-            let store_info = StoreInfo {
-                kv_engine: self.ctx.engines.kv.clone(),
-                raft_engine: self.ctx.engines.raft.clone(),
-                capacity: self.ctx.cfg.capacity.0,
-            };
-            let task = PdTask::StoreHeartbeat {
-                stats,
-                store_info,
-                send_detailed_report: true,
-                dr_autosync_status: self
-                    .ctx
-                    .global_replication_state
-                    .lock()
-                    .unwrap()
-                    .store_dr_autosync_status(),
-            };
-            if let Err(e) = self.ctx.pd_scheduler.schedule(task) {
-                panic!("fail to send detailed report to pd {:?}", e);
-            }
-        }
-        self.fsm.unsafe_recovery_target_commit_index = None;
-        self.fsm.unsafe_recovery_wait_apply_counter = None;
-    }
-
     fn on_unsafe_recovery_wait_apply(&mut self, counter: Arc<AtomicUsize>) {
-        self.fsm.unsafe_recovery_target_commit_index =
-            Some(self.fsm.peer.raft_group.store().commit_index());
-        self.fsm.unsafe_recovery_wait_apply_counter = Some(counter);
-        // If the applied index equals to the commit index, there is nothing to wait for, proceeds
-        // to the next step immediately. If they are not equal, further checks will be performed in
-        // on_apply_res().
-        if self.fsm.stopped
-            || self.fsm.peer.raft_group.store().applied_index()
-                == self.fsm.peer.raft_group.store().commit_index()
-        {
-            self.finish_unsafe_recovery_wait_apply();
+        self.fsm.peer.unsafe_recovery_wait_apply(self.ctx, counter);
```

Comment:
can we remove the `unsafe_recovery_wait_apply` and keep the logic here?

Reply:
Do you mean keep the state variable inside `raftstore/src/store/peer.rs`, but have the logic inside `raftstore/src/store/fsm/peer.rs`? That will be odd, won't it? Isn't it better to have the variables and the functions manipulating those variables close to each other?

Reply:
I agree the boundary between `store/peer.rs` and `store/fsm/peer.rs` is not cleared. We visit states recorded in `store/peer.rs` directly in `store/fsm/peer.rs`, See other `on_xxxx` in the pattern `self.fsm.peer.xxx`. 

Given that, `on_unsafe_recovery_wait_apply` with a  `unsafe_recovery_wait_apply` is verbose.